src/shared/watchdog.c

```diff
@@ -19,26 +19,34 @@ static char *watchdog_device = NULL;
 static usec_t watchdog_timeout = USEC_INFINITY;
 static usec_t watchdog_last_ping = USEC_INFINITY;
 
+/* Starting from kernel version 4.5, the maximum allowable watchdog timeout is
+ * UINT_MAX/1000U seconds (since internal calculations are done in milliseconds).
+ */
+#define WATCHDOG_TIMEOUT_MAX_SEC (UINT_MAX/1000U)
+
+static unsigned saturated_usec_to_sec(usec_t val) {
+        usec_t t = DIV_ROUND_UP(val, USEC_PER_SEC);
+        return (unsigned)(t >= WATCHDOG_TIMEOUT_MAX_SEC ? WATCHDOG_TIMEOUT_MAX_SEC : t); /* Saturate to watchdog max */
+}
+
 static int update_timeout(void) {
         if (watchdog_fd < 0)
                 return 0;
         if (watchdog_timeout == USEC_INFINITY)
                 return 0;
 
         if (watchdog_timeout == 0) {
-                int flags;
+                unsigned flags;
 
                 flags = WDIOS_DISABLECARD;
                 if (ioctl(watchdog_fd, WDIOC_SETOPTIONS, &flags) < 0)
```

Comment:
kernel docs (Documentation/watchdog/watchdog-api.rst) and ioctl definition both suggest "int" is the right type:

```c
define   WDIOC_SETOPTIONS        _IOR(WATCHDOG_IOCTL_BASE, 4, int)
```

The selftest the kernel sources have also used unsigned: tools/testing/selftests/watchdog/watchdog-test.c.

All the copy_from_user() calls that copy the field from userspace into kernel also all specify int not unsigned int... Including the one in drivers/watchdog/watchdog_dev.c which supposedly implements the new world order?

I mean, I appreciate if the kernel peeps get their type system a bit more in order, but here we call the userpsace APIs the kernel provides, not the internal code, hence we should follow what is declared + documented + implemented for that, and what it then does internally doesn't really matter for us here

Reply:
when the generic implementation of WDIOC_SETOPTIONS propagates things to watchdog_set_timeout it passes a signed int to an unsigned int argument. That's icky I must say, but it's internal to the kernel, so we don't really need to care about that in userspace.